internal/provider/resource_gitlab_user_gpgkey.go

```diff
@@ -0,0 +1,187 @@
+			"user_id": {
+				Description: "The ID of the user to add the GPG key to. If this field is omitted, this resource manages a GPG key for the current user. Otherwise, this resource manages a GPG key for the speicifed user, and an admin token is required.",
+				Type:        schema.TypeInt,
+				ForceNew:    true,
+				Optional:    true,
+			},
+			"key": {
+				Description: "The armored GPG public key.",
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Required:    true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					strippedOld := strings.TrimSpace(old)
+					strippedNew := strings.TrimSpace(new)
+					return strippedOld == strippedNew
+				},
+			},
+			"key_id": {
+				Description: "The ID of the GPG key.",
+				Type:        schema.TypeInt,
+				Computed:    true,
+			},
+			"created_at": {
+				Description: "The time when this key was created in GitLab.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+		},
+	}
+})
+
+func resourceGitlabUserGPGKeyCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	options := &gitlab.AddGPGKeyOptions{
+		Key:   gitlab.String(strings.TrimSpace(d.Get("key").(string))),
+	}
+
+	var key *gitlab.GPGKey
+	var err error
+	userID, userIDOk := d.GetOk("user_id")
+	if userIDOk {
+		isAdmin, isAdminErr := isCurrentUserAdmin(ctx, client)
+		if isAdminErr != nil {
+			return diag.Errorf("failed to check if user is admin for configuring GPG keys for a user")
+		}
+		if !isAdmin {
+			return diag.Errorf("current user needs to be admin for configuring GPG keys for a user")
+		}
+		key, _, err = client.Users.AddGPGKeyForUser(userID.(int), options, gitlab.WithContext(ctx))
+	} else {
+		key, _, err = client.Users.AddGPGKey(options, gitlab.WithContext(ctx))
+	}
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	keyIDForID := fmt.Sprintf("%d", key.ID)
+	if userIDOk {
+		userIDForID := fmt.Sprintf("%d", userID)
+		d.SetId(buildTwoPartID(&userIDForID, &keyIDForID))
+	} else {
+		d.SetId(keyIDForID)
+	}
+	return resourceGitlabUserGPGKeyRead(ctx, d, meta)
+}
+
+func resourceGitlabUserGPGKeyRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	userID, keyID, err := resourceGitlabUserGPGKeyParseID(d.Id())
+	if err != nil {
+		return diag.Errorf("unable to parse user GPG key resource id: %s: %v", d.Id(), err)
+	}
+
+	var keys []*gitlab.GPGKey
+	if userID != 0 {
+		keys, _, err = client.Users.ListGPGKeysForUser(userID, gitlab.WithContext(ctx))
+	} else {
+		keys, _, err = client.Users.ListGPGKeys(gitlab.WithContext(ctx))
+	}
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	var key *gitlab.GPGKey
+	for _, k := range keys {
+		if k.ID == keyID {
+			key = k
+			break
+		}
+	}
+
+	if key == nil {
+		log.Printf("Could not find GPG key %d for user %d", keyID, userID)
+		d.SetId("")
+		return nil
+	}
```

Comment:
Any particular reason for not using `client.Users.GetGPGKey` and `client.users.GetGPGKeyForUser` ?

This would simplify the code quite a bit (and we wouldn't need to implement pagination logic, which is a problem in the current implementation ;))

Something along the lines of this: (didn't actually test this)

```suggestion
	var key *gitlab.GPGKey
	if userID != 0 {
		key, _, err = client.Users.GetGPGKeyForUser(userID, keyID, gitlab.WithContext(ctx))
	} else {
		key, _, err = client.Users.GetGPGKey(keyID, gitlab.WithContext(ctx))
	}
	if err != nil {
		if is404(err) {
			log.Printf("Could not find GPG key %d for user %d, removing from state", keyID, userID)
			d.SetId("")
			return nil
		}
		return diag.FromErr(err)
	}
```

Reply:
> Any particular reason for not using `client.Users.GetGPGKey` and `client.users.GetGPGKeyForUser` ?

Because I copy the code from `resource_gitlab_user_sshkey.go` :joy: 

Your suggested fix "almost" works, many thanks! The only missing bit is another bug in go-gitlab - https://github.com/xanzy/go-gitlab/pull/1512

Reply:
> Because I copy the code from `resource_gitlab_user_sshkey.go` 😂

Yes, I've only recently added this to GitLab itself: https://gitlab.com/gitlab-org/gitlab/-/merge_requests/81790

> Your suggested fix "almost" works, many thanks! The only missing bit is another bug in go-gitlab - https://github.com/xanzy/go-gitlab/pull/1512

Awesome! Thanks for creating to PR over at go-gitlab 🎉